spec/firebase_app_distribution_action_spec.rb

```diff
@@ -1,141 +1,131 @@
 describe Fastlane::Actions::FirebaseAppDistributionAction do
   let(:fake_file) { StringIO.new }
+  let(:fake_connection) { double("Connection") }
+  let(:fake_binary) { double("Binary") }
 
-  before do
-    allow(Tempfile).to receive(:new).and_return(fake_file)
-    allow(fake_file).to receive(:unlink)
-    allow(fake_file).to receive(:path).and_return("/tmp/string")
+  before(:each) do
+    allow(Fastlane::Actions::FirebaseAppDistributionAction).to receive(:connection).and_return(fake_connection)
+    allow(fake_binary).to receive(:read).and_return("Hello World")
+    allow(File).to receive(:open).and_return(fake_binary)
   end
 
-  describe '#run' do
-    it 'shells out to firebase' do
-      expect(Fastlane::Actions).to receive(:sh_control_output).with("/tmp/fake-firebase-cli appdistribution:distribute /tmp/FakeApp.ipa --app abc:123 --testers-file /tmp/testers.txt --release-notes-file /tmp/release_notes.txt --token fake-token --debug", { print_command: false, print_command_output: true })
-      params = {
-        app: "abc:123",
-        ipa_path: "/tmp/FakeApp.ipa",
-        firebase_cli_path: "/tmp/fake-firebase-cli",
-        testers_file: "/tmp/testers.txt",
-        release_notes_file: "/tmp/release_notes.txt",
-        firebase_cli_token: "fake-token",
-        debug: true
-      }
-      Fastlane::Actions::FirebaseAppDistributionAction.run(params)
-    end
-
-    it 'removes trailing newlines from firebase_cli_path' do
-      expect(Fastlane::Actions).to receive(:sh_control_output).with("/tmp/fake-firebase-cli appdistribution:distribute /tmp/FakeApp.ipa --app abc:123 --testers-file /tmp/testers.txt --release-notes-file /tmp/release_notes.txt --token fake-token --debug", { print_command: false, print_command_output: true })
-      params = {
-          app: "abc:123",
-          ipa_path: "/tmp/FakeApp.ipa",
-          firebase_cli_path: "/tmp/fake-firebase-cli\n",
-          testers_file: "/tmp/testers.txt",
-          release_notes_file: "/tmp/release_notes.txt",
-          firebase_cli_token: "fake-token",
-          debug: true
-      }
-      Fastlane::Actions::FirebaseAppDistributionAction.run(params)
-    end
-  end
-
-  describe "flag helpers" do
-    let(:action) { Fastlane::Actions::FirebaseAppDistributionAction }
-
-    describe "flag_value_if_supplied" do
-      it "returns flag and param value when it exists" do
-        params = {
-            firebase_cli_token: 'fake-token'
-        }
-        expect(action.flag_value_if_supplied('--token', :firebase_cli_token, params)).to eq("--token fake-token")
-      end
-
-      it "returns nil if param does not exist" do
-        params = {}
-        expect(action.flag_value_if_supplied('--token', :firebåse_cli_token, params)).to be_nil
+  describe '#get_upload_token' do
+    context 'when testing with valid parameters' do
```

Comment:
You can probably drop the `when testing` from the context message here.

But if possible I recommend avoiding using extra nesting with the `context` block because it can easily provide a way for you to accidentally introduce a [mystery guest](https://groups.google.com/a/google.com/forum/#!msg/testing-on-the-toilet/hci1Y-JM5g8/kheVY9WOBQAJ) by providing the perfect spot to start putting extra fixture code which becomes confusing.  If you can make an `it` clause that's terse enough, consider removing this extra `context` indentation entirely.

Reply:
I'm a little confused by the mystery guest concept and unfortunately don't have access to the corp links. Are mystery guests when you have code say within the "when testing with valid parameters" portion that may affect the individual tests within that context?

Reply:
Updated the reference link to something publically shared.  Basically using context/let can provide an opportunity for 'surprises' to creep into your specs because the fixture setup happens far away from the actual test code.